_posts/2018-11-25-rafting-for-fun-and-profit.md

```diff
@@ -0,0 +1,442 @@
+            this.state.alreadyVotedInThisTerm = true;
+
+            const votes = await this.getVotes();
+            const positive = this.countPositiveVotes(votes);
+            logger.log('received positive votes', positive);
+            this.state.votesGranted = positive;
+
+            if (positive > (this.state.peers.size / 2)) {
+                clearTimeout(this.leaderTimeout);
+                this.transitionToLeader();
+            } else {
+                this.resetLeaderTimeout(); // (1)
+            }
+        }
+    }
+
+    resetLeaderTimeout() {
+        clearTimeout(this.leaderTimeout);
+        this.leaderTimeout = setTimeout(this.transitionToCandidate.bind(this), config.leaderTimeout);
+    }
+
+    // ...
+}
+```
+
+But it's kind of "janky"[*](http://jankfree.org/)...
+Sometimes the test completes in milliseconds,
+sometimes it runs for way longer...
+To understand why I had to refresh my memory about event loop
+and conditions for a node process to exit.
+
+Every asynchronous *thing* is going to be queued up in the event loop
+to be processed at the later time.
+A node process will not exit if it has anything left to do in the event loop.
+I made sure that the event loop was constantly occupied
+by clearing a timeout an starting a new one every time I called `resetLeaderTimeout()`.
+
+The situation looks like this when `transitionToCandidate()` is called:
+
+```
+Stack: [transitionToCandidate()]
+Event loop: []
+```
+
+And by the time it exits it's on the event loop (1)
+
+```
+Stack: []
+Event loop: [transitionToCandidate()]
+```
+
+Ok, so how to fix it?
+There is a function that can be called on a
+`timeout`,
+`immediate`
+and `subprocess`
+called [unref](https://nodejs.org/api/timers.html#timers_timeout_unref)
+which makes "object not require the Node.js event loop to remain active".
+
+Alright, we fixed that one, but it's not ideal.
+Our tests take more time than they should and it's because we use real timers.
+
+```
+ PASS  state-machine/state-machine.test.js
+  ✓ should start with initial state (4ms)
+  ✓ should not transition to leader when can not get majority (38ms)
+  ✓ should transition to candidate role when it did not receive a heartbeat (17ms)
+  ✓ should send heartbeats when leader (1ms)
+  ✓ should become follower when it gets heartbeat with higher term
+  ✓ should become candidate when it does not get heartbeat from a leader for a long time (16ms)
+  ✓ should start a second voting when first vote is unsuccessful (16ms)
+  ✓ should cast a vote when not voted in this term (1ms)
+  ✓ should not cast a vote when voted in this term
+  ✓ should elect a leader (50ms)
+  starting voting
+    ✓ should transition from candidate to leader when recived majority of votes (1ms)
+    ✓ should not transition from candidate to leader when recived less than majority of votes
+```
+
+Sum: ~144 ms
+
+What can we do about it?
+[Jest](https://jestjs.io/docs/en/timer-mocks) allows mocking timers and
+we're going to use it.
```

Comment:
Nie znam narzędzia, ale może dałoby się jakoś wyciągnąć timer na zewnątrz i po prostu mockować? Temat oszukiwania czasu, może chyba spowodować lekką wojenkę w komentarzach. Przynajmniej wspomniałbym inne opcje, bo trochę dogmatycznie "we're going to use it" ;)